pubsub/gossipsub/gossipsub-v1.2.md

```diff
@@ -0,0 +1,102 @@
+
+On the other side a node maintains per-peer `dont_send_message_ids` set. Upon receiving `DONTSEND` from 
+a peer the `messageId` is added to the `dont_send_message_ids` set. 
+When later relaying the `messageId` message to the mesh the peers found in `dont_send_message_ids` could be skipped. 
+
+Old entries from `dont_send_message_ids` could be pruned during heartbeat processing. 
+The prune strategy is outside of the spec scope and can be decided by implementations.
+
+`DONTSEND` message is supposed to be _optional_ for both receiver and sender. I.e. the sender may or may not utilize 
+this message. The receiver in turn may ignore `DONTSEND`: sending a message after the corresponding `DONTSEND` 
+should not be penalized.    
+
+The `DONTSEND` may have negative effect on small messages as it may increase the overall traffic and CPU load.
+Thus it is better to utilize `DONTSEND` for messages of a larger size.
+The exact policy of `DONTSEND` appliance is outside of the spec scope. Every implementation may choose whatever 
+is more appropriate for it. Possible options are either choose a message size threshold and broadcast `DONTSEND`
+on per message basis when the size is exceeded or just use `DONTSEND` for all messages on selected topics.
+
+To prevent DoS the number of `DONTSEND` control messages is limited to `max_dontsend_messages` per heartbeat  
+
+### Relying on `IHAVE`s
+
+Another potential additional strategy could be as follows. If a node receives `IHAVE` (from one or more peers)
+before the message is appeared in the mesh the node may request the message with `IWANT` and notify all mesh 
+peers that it don't want that message from them. 
+
+### Sending `IHAVE` to mesh peers who choked that particular message
+
+Reasonable addition to the later scenario would be to _immediately_ send `IHAVE` instead of a full message
+to those mesh peers who reported `DONTSEND`. That would notify mesh peers that the node has this message 
+and they could request it from you in case their `IWANT` requests fail in the previous scenario 
+
+### Cancelling `IWANT`
+
+If a node requested a message via `IWANT` and then occasionally receives the message from other peer it may 
+try to cancel its `IWANT` requests with the corresponding `DONTSEND` message. It may work in cases when a
+peer delays/queues `IWANT` requests and the `IWANT` request would be removed from the queue if not processed yet
+
+## Protobuf Extension
+
+The protobuf messages are identical to those specified in the [gossipsub v1.0.0
+specification](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md)
+with the following  control message modifications:
+
+```protobuf
+message RPC {
+ // ... see definition in the gossipsub specification
+}
+
+message ControlMessage {
+    // messages from v1.0
+    repeated ControlDontSend dontSend = 5;
+}
+
+message ControlDontSend {
+    required bytes messageID = 1;
```

Comment:
Just for discussion, what would be the pros/cons of also including the topic?
I think it's the first time that we reference messages only by their id on the wire, so needs some consideration

I guess the only con of include the topic is more bandwidth usage

Reply:
If no one is opposed or has any ideas on usage scenarios I would keep it without optional topic field to be more explicit

Reply:
> If no one is opposed or has any ideas on usage scenarios I would keep it without optional topic field to be more explicit

The usage can be https://github.com/libp2p/specs/pull/548#discussion_r1534137418